OpenSim/Common/Component.cpp

```diff
@@ -865,11 +880,29 @@ SimTK::Vector Component::
     getStateVariableValues(const SimTK::State& state) const
 {
     int nsv = getNumStateVariables();
-    Array<std::string> names = getStateVariableNames();
+    // Consider the list of all StateVariables to be valid if all of 
+    // the following are true:
+    // 1. Component is up-to-date with its Properties
+    // 2. a System has been associated with the list of StateVariables
+    // 3. The list of all StateVariables is correctly sized (initialized)
+    // 4. The System associated with the StateVariables is the current System
+    bool valid = isObjectUpToDateWithProperties() &&   // 1.
+                _statesAssociatedSystem.empty()  &&     // 2.
+                _allStateVariables.size() == nsv &&     // 3.
+                getSystem().isSameSystem(_statesAssociatedSystem.getRef());// 4.
+    // if the StateVariables are invalid (see above) rebuild the list 
+    if (!valid) {
```

Comment:
Hm, okay. I thought it might be useful to add a Debug mode–only assert here, just in case there's some other condition the needs to be checked in the future (and assuming the existing tests don't provide sufficiently detailed diagnostic information). I don't know the new API well enough to know whether these four conditions constitute a necessary and sufficient set, but the comments didn't immediately convince me that there isn't something missing. One specific question, though: why isn't `_statesAssociatedSystem.empty()` negated? Don't you want to check that it is _not_ empty? :water_buffalo: